src/uct/tcp/tcp_ep.c

```diff
@@ -788,6 +893,95 @@ ssize_t uct_tcp_ep_am_bcopy(uct_ep_h uct_ep, uint8_t am_id,
+{
+    uct_tcp_ep_t *ep       = ucs_derived_of(uct_ep, uct_tcp_ep_t);
+    uct_tcp_iface_t *iface = ucs_derived_of(uct_ep->iface, uct_tcp_iface_t);
+    uct_tcp_am_hdr_t *hdr  = NULL;
+    uct_tcp_ep_zcopy_ctx_t *ctx;
+    struct iovec *ctx_iov;
+    ucs_status_t status;
+
+    UCT_CHECK_IOV_SIZE(iovcnt, iface->config.zcopy.max_iov,
+                       "uct_tcp_ep_am_zcopy");
+    UCT_CHECK_LENGTH(header_length, 0, iface->config.zcopy.max_hdr,
+                     "am_zcopy header");
+    UCT_CHECK_LENGTH(header_length + uct_iov_total_length(iov, iovcnt), 0,
+                     iface->config.rx_seg_size - sizeof(uct_tcp_am_hdr_t),
+                     "am_zcopy");
+
+    status = uct_tcp_ep_am_prepare(iface, ep, am_id, &hdr);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ucs_assertv(hdr != NULL, "ep=%p", ep);
+
+    ctx          = ucs_derived_of(hdr, uct_tcp_ep_zcopy_ctx_t);
+    ctx->iov_cnt = 0;
+    ctx_iov      = (struct iovec*)ctx->iov;
+
+    ctx_iov[ctx->iov_cnt].iov_base = hdr;
+    ctx_iov[ctx->iov_cnt].iov_len  = sizeof(*hdr);
+    ctx->iov_cnt++;
+
+    if (header_length != 0) {
+        ucs_assert(header != NULL);
+        ctx_iov[ctx->iov_cnt].iov_base = (void*)header;
+        ctx_iov[ctx->iov_cnt].iov_len  = header_length;
+        ctx->iov_cnt++;
+    }
+
+    ctx->iov_cnt += uct_tcp_ep_iovec_fill_iov(&ctx_iov[ctx->iov_cnt], iov,
+                                              iovcnt, &ep->tx.length);
+    hdr->length   = ep->tx.length + header_length;
+    ep->tx.length = hdr->length   + sizeof(*hdr);
+
+    ucs_assertv(ep->tx.length <= iface->config.rx_seg_size, "ep=%p", ep);
+
+    uct_iface_trace_am(&iface->super, UCT_AM_TRACE_TYPE_SEND, hdr->am_id,
+                       header, hdr->length, "SEND %p fd %d", ep, ep->fd);
+
+    iface->outstanding += ep->tx.length;
+
+    status = ucs_socket_sendv_nb(ep->fd, ctx_iov, ctx->iov_cnt,
+                                 &ep->tx.offset, NULL, NULL);
+
+    ucs_trace_data("tcp_ep %p: sendv %zu bytes", ep, ep->tx.offset);
+
+    iface->outstanding -= ep->tx.offset;
+    if ((ep->tx.offset != ep->tx.length) &&
+        ((status == UCS_OK) || (status == UCS_ERR_NO_PROGRESS))) {
+        ctx->comp     = comp;
+        ep->ctx_caps |= UCS_BIT(UCT_TCP_EP_CTX_TYPE_ZCOPY_TX);
+
+        if ((header_length != 0) &&
+            /* check whether a user's header was sent or not */
+            (ep->tx.offset < (sizeof(*hdr) + header_length))) {
+            ucs_assert(header_length <= iface->config.zcopy.max_hdr);
+            /* if the user's header wasn't sent completely, copy it to the EP
+             * TX buffer (after Zcopy context and IOVs) for retransmission */
+            ctx_iov[1].iov_base = ep->tx.buf + iface->config.zcopy.hdr_offset;
+            memcpy(ctx_iov[1].iov_base, header, header_length);
+        }
+
+        ctx->iov_index  = 0;
+        ctx->iov_offset = 0;
+
+        uct_tcp_ep_zcopy_ctx_consume_iov(ctx, ep->tx.offset);
```

Comment:
maybe uct_tcp_ep_zcopy_ctx_consume_iov should be called directly from ucs_socket_sendv_nb()?

Reply:
you're right, that is in another layer.